core/src/banking_stage/worker.rs

```diff
@@ -0,0 +1,626 @@
+        forwarder::Forwarder,
+        scheduler_messages::{ConsumeWork, FinishedConsumeWork, FinishedForwardWork, ForwardWork},
+        ForwardOption,
+    },
+    crossbeam_channel::{select, Receiver, RecvError, SendError, Sender},
+    solana_poh::leader_bank_notifier::LeaderBankNotifier,
+    solana_runtime::bank::Bank,
+    std::{sync::Arc, time::Duration},
+    thiserror::Error,
+};
+
+#[derive(Debug, Error)]
+pub enum WorkerError {
+    #[error("Failed to receive work from scheduler: {0}")]
+    Recv(RecvError),
+    #[error("Failed to send finalized consume work to scheduler: {0}")]
+    ConsumedSend(SendError<FinishedConsumeWork>),
+    #[error("Failed to send finalized forward work to scheduler: {0}")]
+    ForwardedSend(SendError<FinishedForwardWork>),
+}
+
+impl From<RecvError> for WorkerError {
+    fn from(err: RecvError) -> Self {
+        Self::Recv(err)
+    }
+}
+
+impl From<SendError<FinishedConsumeWork>> for WorkerError {
+    fn from(err: SendError<FinishedConsumeWork>) -> Self {
+        Self::ConsumedSend(err)
+    }
+}
+
+impl From<SendError<FinishedForwardWork>> for WorkerError {
+    fn from(err: SendError<FinishedForwardWork>) -> Self {
+        Self::ForwardedSend(err)
+    }
+}
+
+pub(crate) struct Worker {
+    consume_receiver: Receiver<ConsumeWork>,
+    consumer: Consumer,
+    consumed_sender: Sender<FinishedConsumeWork>,
+
+    forward_receiver: Receiver<ForwardWork>,
+    forward_option: ForwardOption,
+    forwarder: Forwarder,
+    forwarded_sender: Sender<FinishedForwardWork>,
+
+    leader_bank_notifier: Arc<LeaderBankNotifier>,
+}
+
+impl Worker {
+    pub fn new(
+        consume_receiver: Receiver<ConsumeWork>,
+        consumer: Consumer,
+        consumed_sender: Sender<FinishedConsumeWork>,
+        forward_receiver: Receiver<ForwardWork>,
+        forward_option: ForwardOption,
+        forwarder: Forwarder,
+        forwarded_sender: Sender<FinishedForwardWork>,
+        leader_bank_notifier: Arc<LeaderBankNotifier>,
+    ) -> Self {
+        Self {
+            consume_receiver,
+            consumer,
+            consumed_sender,
+            forward_receiver,
+            forward_option,
+            forwarder,
+            forwarded_sender,
+            leader_bank_notifier,
+        }
+    }
+
+    pub fn run(self) -> Result<(), WorkerError> {
+        loop {
+            select! {
+                recv(self.consume_receiver) -> consume_work => {
+                    self.consume_loop(consume_work?)?;
+                },
+                recv(self.forward_receiver) -> forward_work => {
+                    self.forward_loop(forward_work?)?;
+                },
+            }
+        }
+    }
+
+    fn consume_loop(&self, consume_work: ConsumeWork) -> Result<(), WorkerError> {
+        let Some(mut bank) = self.get_consume_bank() else {
+            return self.retry_drain(consume_work);
+        };
+
+        for work in std::iter::once(consume_work).chain(self.consume_receiver.try_iter()) {
```

Comment:
seems `once(...).chain(..try_iter())` is happening multiple times. how about factoring these into a small fn?

Reply:
77eb8338047d0f23b1b3ffec8a9bd3d7742b30be